internal/api/core/user.go

```diff
@@ -0,0 +1,171 @@
+package core
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/edulinq/autograder/internal/db"
+	"github.com/edulinq/autograder/internal/log"
+	"github.com/edulinq/autograder/internal/model"
+)
+
+type UserInfoType string
+
+const (
+	ServerUserInfoType = "ServerType"
+	CourseUserInfoType = "CourseType"
+)
+
+// These fields must be embedded into any API-safe representation of a user.
+type UserInfo struct {
+	Type  UserInfoType `json:"type"`
+	Email string       `json:"email"`
+	Name  string       `json:"name"`
+}
+
+// An API-safe representation of a server user.
+type ServerUserInfo struct {
+	UserInfo
+	Role    model.ServerUserRole      `json:"role"`
+	Courses map[string]EnrollmentInfo `json:"courses"`
+}
+
+type EnrollmentInfo struct {
+	CourseID   string               `json:"id"`
+	CourseName string               `json:"name"`
+	Role       model.CourseUserRole `json:"role"`
+}
+
+// An API-safe representation of a course user.
+type CourseUserInfo struct {
+	UserInfo
+	Role  model.CourseUserRole `json:"role"`
+	LMSID string               `json:"lms-id"`
+}
+
+func NewServerUserInfo(user *model.ServerUser) (*ServerUserInfo, error) {
+	info := &ServerUserInfo{
+		UserInfo: UserInfo{
+			Type:  ServerUserInfoType,
+			Email: user.Email,
+			Name:  user.GetDisplayName(),
+		},
+		Role:    user.Role,
+		Courses: make(map[string]EnrollmentInfo, len(user.CourseInfo)),
+	}
+
+	for courseID, courseInfo := range user.CourseInfo {
+		course, err := db.GetCourse(courseID)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get course '%s' for user '%s'.", courseID, user.Email)
+		}
+
+		info.Courses[course.GetID()] = EnrollmentInfo{
+			CourseID:   course.GetID(),
+			CourseName: course.GetName(),
+			Role:       courseInfo.Role,
+		}
+	}
+
+	return info, nil
+}
+
+func MustNewServerUserInfo(user *model.ServerUser) *ServerUserInfo {
+	info, err := NewServerUserInfo(user)
+	if err != nil {
+		log.Fatal("Failed to convert server user to API info.", err, user)
```

Comment:
Use log.NewAttr("user",  user) or something to list the user instead to maintain consistency.